File: app/src/jvx/java/art/coded/wireframe/view/fragment/PagingFragment.java
package art.coded.wireframe.view.fragment;

import android.app.Activity;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import art.coded.wireframe.model.entity.Element;
import art.coded.wireframe.model.entity.ElementComparator;
import art.coded.wireframe.databinding.FragmentPagingBinding;
import art.coded.wireframe.view.adapter.PagingAdapter;
import art.coded.wireframe.viewmodel.PagingViewModel;

public class PagingFragment extends Fragment {

    static final String LOG_TAG = PagingFragment.class.getSimpleName();

    private FragmentPagingBinding binding;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentPagingBinding.inflate(inflater, container, false);
        View root = binding.getRoot();

        Activity activity = getActivity();


        final RecyclerView recyclerView = binding.rvPaging;
        PagingAdapter pagingAdapter = new PagingAdapter(new ElementComparator(), requireActivity());
        recyclerView.setAdapter(pagingAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(activity));

        PagingViewModel pagingViewModel =
                new ViewModelProvider(this).get(PagingViewModel.class);
        pagingViewModel.loadData(requireActivity().getApplication());
//        List<Element> elements = pagingViewModel.elementList().getValue();

        pagingViewModel.elementList().observe(requireActivity(), pagingAdapter::submitList);

        return root;
    }

    @Override public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/ktx/java/art/coded/wireframe/view/fragment/PagingFragment.kt
package art.coded.wireframe.view.fragment

import androidx.paging.PagedList
import android.app.Activity
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import androidx.lifecycle.ViewModelProvider
import art.coded.wireframe.viewmodel.PagingViewModel
import art.coded.wireframe.view.adapter.PagingAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import android.view.View
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import art.coded.wireframe.databinding.FragmentPagingBinding
import art.coded.wireframe.model.entity.Element
import art.coded.wireframe.model.entity.ElementComparator

private val LOG_TAG = PagingFragment::class.java.simpleName

class PagingFragment : Fragment() {
    private var binding: FragmentPagingBinding? = null
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentPagingBinding.inflate(inflater, container, false)
        val root: View = binding!!.root
        val activity: Activity? = activity
        val recyclerView = binding!!.rvPaging
        val pagingAdapter = PagingAdapter(ElementComparator(), requireActivity())
        recyclerView.adapter = pagingAdapter
        recyclerView.layoutManager = LinearLayoutManager(activity)
        pagingAdapter.stateRestorationPolicy = RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY;
        val pagingViewModel = ViewModelProvider(this).get(PagingViewModel::class.java)
        pagingViewModel.loadData(requireContext())
        //        List<Element> elements = pagingViewModel.elementList().getValue();
        pagingViewModel.elementList()
            .observe(requireActivity()) { pagedList: PagedList<Element> ->
                pagingAdapter.submitList(
                    pagedList
                )
            }
        return root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}